Add timer trigger for amp-analytics

breauxc opened the pull request:
This is an initial PR to address issue #1489. I'm putting up a light implementation with the hope of soliciting feedback and confirming direction.

The following code implements the timer trigger on the following analytics tag:
```
    <amp-analytics>
    <script type="application/json">
    {
      "requests": {
        "test-ping": "https://my-analytics.com/ping?title=${title}&acct=${account}"
      },
      "vars": {
        "title": "A page that sends a ping",
        "account": "12345"
      },
      "triggers": {
        "TRIGGER_NAME": {
          "on": "**timer**",
          "timer-spec": { "**interval**": 3 /* seconds */ },
          "request": "test-ping"
        }
      }
    }
    </script>
    </amp-analytics>
```
We put a minimum on the length of the timer interval at half a second because we don't want to encourage a process that is constantly firing.

While we were thinking through this implementation, we wished to get feedback on a number of possible refinements to this approach (listed in roughly decreasing order of importance):

1. Handling the trailing ping. We would like to be able to capture data between the most recent timer event and the close of a user session. This becomes increasingly important as the length of the timer interval increases. We agree with the feedback to avoid use of the unload and beforeunload events to accomplish this, but we would like to open the conversation about how best to proceed.
2. Max Session Length.  The initial specification for the timer-spec included a max-count parameter. We think that a more intuitive/useful specification would be max-session-length. In either case, we recognize that we probably do not want an idle session to ping forever. I have additional code implementing this that I am happy to add to this PR.
3. Session Backoff.  We have found it useful to ping progressively less frequently when the user is not engaged.
4. Sequence Number. It can be useful for sessionization to expose the sequence number of a user session (e.g., &seq-no=3 to indicate a ping was the third ping in the session). This is a very easy, if perhaps non-essential, extension.
5. Error Handling. We attach onerror functions to pings so that we can halt the timer on a specified client error code.

cramforce commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -18,6 +18,9 @@ import {getService} from '../../../src/service';
 import {Observable} from '../../../src/observable';
 
+
+const MIN_TIMER_INTERVAL_ = 0.5;
```
Please put unit into variable name.

cramforce commented:
Looks great! Assigning to @avimehta

avimehta commented:
thanks @breauxc !

The commit looks good so far. 

>  Handling the trailing ping

It would have been nice to trigger pings based on other variables like active-time etc. Then you can send data when enough of it is there to send. Unfortunately that part is not built I don't really have other ideas. maybe @cramforce has some ideas?

> Max Session Length

`max-session-length` works for me. Can you add it to the PR? Since the term `session` is overloaded, would it make sense to call it `max-timer-length`? 

> Session Backoff

Should the session back-off come with some cap? That would reduce the need for trailing ping?

> Sequence Number

Would requestCount variable work for this purpose? I added it recently #1598.

> Error Handling

sounds good. Please add it to the commit.

breauxc commented:
Thanks @cramforce and @avimehta for taking an initial look at this.

I cleaned up the code a bit with your helpful suggestions and added the max-timer-length specification. I agree with you that requestCount should work well for the purposes of Sequence Number.

My suggestion at this point is perhaps to split off the error handling piece into a separate issue with lower priority. I'm happy to describe/implement a process that has worked well for us thus far. However, it seems that Beacon requests are intended to strictly be one-way communication, and don't really support a (non-cumbersome) way to handle errors.

> Trailing ping;
> Timer Backoff
> Should the session back-off come with some cap? That would reduce the need for trailing ping?

We currently use a 15 second interval in our specification. Even without back-off, we'd lose a substantial amount of resolution if we don't capture page/engagement for users that leave before 15 seconds. As an interim solution, we can additionally specify a click trigger (to capture click-thrus), but we still lose data on users who refresh or leave the page. You're right that implementing back-off exacerbates the trailing ping issue.

avimehta commented:
Have not looked into the latest commit. Will probably comment on it tomorrow. For the trailing ping, I have an idea but don't know if it works. You can try it to see if it helps you guys.

The vendor.js file can also contain triggers. If you want, you can add a click trigger with '*' as selector and add the trailing ping for everyone who is using chartbeat. I'll let you guys experiment with it. I am planning on doing that for our latency hits but have not looked into it yet. Please use sendBeacon for this or else hits might not get off of the page in time.

The reason I do not want it(triggers in vendors.js) to be a common practice is because publishers should have full control over what is going out of a page. trailing pings are probably a low-overhead thing and help pubs more than they burden them.

jridgewell commented on extensions/amp-analytics/0.1/instrumentation.js:
```diff
@@ -173,6 +184,43 @@ class InstrumentationService {
+    if (!timerSpec) {
+      console./*OK*/error(this.TAG_, 'Bad timer specification');
+      return false;
+    } else if (!(timerSpec['interval'] >= MIN_TIMER_INTERVAL_SECONDS_)) {
```
Just `<` would be better here.

avimehta commented:
The CL looks good to me. Please add documentation to [amp-analytics.md](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/amp-analytics.md) and some examples to the custom config in [analytics.amp.html](https://github.com/ampproject/amphtml/blob/master/examples/analytics.amp.html). 

Thanks

breauxc commented:
Documentation and examples added.

breauxc commented:
Added a default max-timer-length per @avimehta 's suggestion and squashed commits. Should be good to go, pending final review.

cramforce commented:
Great! Please rebase the branch against master, so we can merge!

avimehta commented:
I'll wait to submit #1767 so that you don't have to rebase and re-merge.

breauxc commented:
@cramforce Rebased. Thanks!

avimehta commented:
This still seems to have some conflicts with HEAD.

cramforce commented:
I manually merged it. Thank you @breauxc!!!

rudygalfi commented:
:+1: Woohoo! Thanks for building this feature. It's a great enhancement to amp-analytics.
